core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -894,7 +899,7 @@ class DAGScheduler(
     submitStage(finalStage)
 
     // If the whole stage has already finished, tell the listener and remove it
-    if (!finalStage.outputLocs.contains(Nil)) {
+    if (finalStage.isAvailable) {
```

Comment:
I believe `!finalStage.outputLocs.contains(Nil)` is just `finalStage.isAvailable`